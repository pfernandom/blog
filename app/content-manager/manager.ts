import { format, parseISO } from 'date-fns'
import fs from 'fs'
import matter from 'gray-matter'
import path, { join } from 'path'

export type Slug = string

export type Markdown = string

export interface FileInfo {
  fileName: string
  filePath: string
  dirPath: string
  subblogPath?: string
}

export interface Post {
  title: string
  date: string
  description: Array<string>
  key_words: Array<string>
  hero_image: string
  hero_image_original?: string
  hero_image_blur: string
  hero_image_alt: string
  hero_width: number
  hero_height: number
  published: boolean
  series: string | null
  social_title: string
  social_subtitle: string
  social_footer: string
  test: boolean
}

export interface SubBlog {
  path: string
  basename: string
  metadata: {
    title: string
    description: string
  }
}

export interface PostInfo {
  slug: Slug
  postPath: string
  frontmatter: Post
  content?: Markdown
}

interface ContentManager {
  fetchAll(subblog: SubBlog): Promise<PostInfo[]>
  fetch(slug: string): Promise<PostInfo | undefined>
}

export class LocalContentManager implements ContentManager {
  contentPath: string

  contentDir: string

  imagesPath: string

  constructor(contentPath: string) {
    this.contentPath = contentPath
    // this.contentDir = path.basename(contentPath)
    this.contentDir = 'app/content/software'
    this.imagesPath = join('/opt_images/', contentPath)
  }

  findSubBlogs(): SubBlog[] {
    const files = fs.readdirSync(this.contentPath)
    return files
      .map((file) => ({ filePath: this.contentPath + '/' + file, slug: file }))
      .filter(
        ({ filePath }) =>
          fs.statSync(filePath).isDirectory() &&
          fs.existsSync(`${filePath}/metadata.json`)
      )
      .map(({ filePath, slug }) => ({
        slug,
        path: filePath,
        basename: this.contentPath,
        metadata: JSON.parse(
          fs.readFileSync(`${filePath}/metadata.json`, 'utf8')
        ),
      }))
  }

  fetch(slug: string): Promise<PostInfo | undefined> {
    throw new Error('Method not implemented.')
  }

  #fetchFromFileInfo(subblog: SubBlog, filePath: FileInfo) {
    const fileContents = fs.readFileSync(filePath.filePath, 'utf8')

    const { data, content } = matter(fileContents)

    if (!data.published) {
      return null
    }

    const date = format(parseISO(data.date), 'MMMM dd, yyyy')
    const {
      title,
      description,
      hero_image: relativeHero,
      hero_image_alt,
      hero_width,
      hero_height,
      published,
      key_words,
      series,
      social_title,
      social_subtitle,
      social_footer,
      test,
    } = data

    const hero_image = relativeHero
      ? path.join(
          this.imagesPath,
          relativeHero.replace(/(png|jpg|jpeg|gif)/gi, 'webp')
        )
      : ''
    const hero_image_blur = relativeHero
      ? path.join(
          this.imagesPath,
          'blur_' +
            relativeHero?.slice(2).replace(/(png|jpg|jpeg|gif)/gi, 'webp')
        )
      : ''

    const hero_image_original = relativeHero
      ? path.join(this.imagesPath, relativeHero)
      : ''

    const frontmatter: Post = {
      title,
      description,
      hero_image,
      hero_image_blur,
      hero_image_original,
      hero_image_alt: hero_image_alt ?? '',
      ...(hero_width && { hero_width }),
      ...(hero_height && { hero_height }),
      date,
      key_words: parseKeyWords(key_words),
      series: series ?? null,
      published,
      social_title: social_title ?? title,
      social_subtitle: social_subtitle ?? description[0],
      social_footer:
        social_footer ?? 'Visit pedromarquez.dev for the full post',
      test: test ?? false,
    }
    const slug = slugToPage(filePath.filePath, new Date(frontmatter.date))
    console.log({ slug })
    return {
      // TODO: Move somewhere else
      slug: slug
        ?.replace(`app/content/software/`, 'blog/')
        ?.replace('/index.mdx', ''),
      postPath: filePath.filePath,
      frontmatter,
      content: content
        .replace(/\]\(.\//g, `](${filePath.filePath}/`)
        .replace(/.gif/g, '.webp'),
    }
  }

  async fetchAll(subblog: SubBlog): Promise<PostInfo[]> {
    const slugs = getAllFiles(this.contentPath, /.mdx?$/gi, [])

    const isProd = process.env.NODE_ENV === 'production'

    const posts: Array<PostInfo> = slugs
      .map((slug) => this.#fetchFromFileInfo(subblog, slug))
      .filter((post) => post != null)
      .filter((post) => {
        if (isProd && post?.frontmatter.test == true) {
          return false
        }
        return true
      })
      .map((post) => post as PostInfo)

    posts.sort(
      (a, b) =>
        new Date(b.frontmatter.date).getTime() -
        new Date(a.frontmatter.date).getTime()
    )
    return posts
  }
}

const getAllFiles = function (
  dirPath: string,
  pattern: string | RegExp,
  arrayOfFiles: Array<FileInfo> | undefined = [],
  relativeToPath = '.',
  isRoot = true
): FileInfo[] {
  const files = fs.readdirSync(dirPath)

  arrayOfFiles = arrayOfFiles || []

  files.forEach(function (file) {
    const regex = RegExp(pattern, 'ig')
    if (fs.statSync(dirPath + '/' + file).isDirectory()) {
      arrayOfFiles = getAllFiles(
        dirPath + '/' + file,
        pattern,
        arrayOfFiles,
        relativeToPath,
        false
      )
    } else if (regex.test(file)) {
      const relativePath = path.relative(
        relativeToPath,
        path.join(dirPath, '/')
      )
      arrayOfFiles.push({
        fileName: file,
        filePath: [relativePath, '/', file].join(''),
        dirPath: relativePath,
      })
    }
  })

  return arrayOfFiles
}

export function slugToPage(path: string, date: Date) {
  const slug = path.replace(/(\d{1,2})_/gi, `${date.getMonth() + 1}/`)
  return slug
}

function parseKeyWords(keywords: string | string[]) {
  if (Array.isArray(keywords)) {
    return keywords.map((keyword) => keyword.trim())
  }
  return (
    keywords
      ?.split(',')
      .map((el) => el.trim())
      .filter((keyword) => keyword.length) ?? []
  )
}
