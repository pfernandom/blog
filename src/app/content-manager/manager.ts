import { CONTENT_PATH } from 'blog_constants'
import fs from 'fs'
import path from 'path'
import parseMDX from './helpers/parse_mdx'
import UrlManager from './url-manager'
import { Post } from 'app/models/interfaces'
export type { Post } from 'app/models/interfaces'

export type Slug = string

export type Markdown = string

export interface FileInfo {
  fileName: string
  filePath: string
  dirPath: string
  subblogPath?: string
}

export interface SubBlog {
  path: string
  basename: string
  metadata: {
    title: string
    description: string
  }
}

export interface PostInfo {
  slug: Slug
  postPath: string
  frontmatter: Post
  content?: Markdown
}

interface ContentManager {
  fetchAll(): Promise<PostInfo[]>
  fetchAllSync(): PostInfo[]
  fetch(slug: string): Promise<PostInfo | undefined>
}

export class LocalContentManager implements ContentManager {
  contentPath: string

  contentDir: string

  urlManager: UrlManager

  constructor() {
    this.contentPath = CONTENT_PATH
    // this.contentDir = path.basename(contentPath)
    this.contentDir = 'app/content/software'

    this.urlManager = new UrlManager()
  }
  fetchAllSync(): PostInfo[] {
    const slugs = getAllFiles(this.contentPath, /.mdx?$/gi, [])

    const isProd = process.env.NODE_ENV === 'production'

    const posts: Array<PostInfo> = slugs
      .map((slug) => this.#fetchFromSubBlogAndFileInfo(slug))
      .filter((post) => post != null)
      .filter((post) => !(isProd && post?.frontmatter.test == true))
      .map((post) => post as PostInfo)

    posts.sort(
      (a, b) =>
        new Date(b.frontmatter.date).getTime() -
        new Date(a.frontmatter.date).getTime()
    )
    return posts
  }

  findSubBlogs(): SubBlog[] {
    const files = fs.readdirSync(this.contentPath)
    return files
      .map((file) => ({ filePath: this.contentPath + '/' + file, slug: file }))
      .filter(
        ({ filePath }) =>
          fs.statSync(filePath).isDirectory() &&
          fs.existsSync(`${filePath}/metadata.json`)
      )
      .map(({ filePath, slug }) => ({
        slug,
        path: filePath,
        basename: this.contentPath,
        metadata: JSON.parse(
          fs.readFileSync(`${filePath}/metadata.json`, 'utf8')
        ),
      }))
  }

  fetch(): Promise<PostInfo | undefined> {
    throw new Error('Method not implemented.')
  }

  #fetchFromSubBlogAndFileInfo(filePath: FileInfo) {
    const fileContents = fs.readFileSync(filePath.filePath, 'utf8')

    const imagesPath = this.urlManager.getImagesDir(filePath.dirPath)

    const { postData: post, content } = parseMDX({
      fileContents,
      imagesPath,
    })

    if (!post.published) {
      return null
    }

    const subblog_slug = this.urlManager.getSubBlogName(filePath.filePath)
    post.subblog_slug = subblog_slug

    const slug = this.urlManager.getSlugFromFilePath(filePath.filePath, post)

    return {
      slug,
      postPath: filePath.filePath,
      frontmatter: post,
      content,
    }
  }

  async fetchAll(): Promise<PostInfo[]> {
    const slugs = getAllFiles(this.contentPath, /.mdx?$/gi, [])

    const isProd = process.env.NODE_ENV === 'production'

    const posts: Array<PostInfo> = slugs
      .map((slug) => this.#fetchFromSubBlogAndFileInfo(slug))
      .filter((post) => post != null)
      .filter((post) => !(isProd && post?.frontmatter.test == true))
      .map((post) => post as PostInfo)

    posts.sort(
      (a, b) =>
        new Date(b.frontmatter.date).getTime() -
        new Date(a.frontmatter.date).getTime()
    )
    return posts
  }
}

const getAllFiles = function (
  dirPath: string,
  pattern: string | RegExp,
  arrayOfFiles: Array<FileInfo> | undefined = [],
  relativeToPath = '.'
): FileInfo[] {
  const files = fs.readdirSync(dirPath)

  arrayOfFiles = arrayOfFiles || []

  files.forEach(function (file) {
    const regex = RegExp(pattern, 'ig')
    if (fs.statSync(dirPath + '/' + file).isDirectory()) {
      arrayOfFiles = getAllFiles(
        dirPath + '/' + file,
        pattern,
        arrayOfFiles,
        relativeToPath
      )
    } else if (regex.test(file)) {
      const relativePath = path.relative(
        relativeToPath,
        path.join(dirPath, '/')
      )
      arrayOfFiles.push({
        fileName: file,
        filePath: [relativePath, '/', file].join(''),
        dirPath: relativePath,
      })
    }
  })

  return arrayOfFiles
}
