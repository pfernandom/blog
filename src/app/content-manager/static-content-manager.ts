import { globSync } from 'glob'
import fs from 'node:fs'
import path from 'path'
// import { reactPlugin, reactOutputPlugin } from './unified'
import { fromFrontMatter } from './unified/frontmatter'
import { parseMarkdown } from './unified/parse'
// import type {IntrinsicElements} from 'react'

export class StaticContentManager {
  getStaticPosts() {
    return globSync(path.join('content/**/*.md'))
      .map((p) => ({ path: p, content: fs.readFileSync(p, 'utf8') }))
      .map(({ path, content }) => ({
        path,
        content,
        data: fromFrontMatter(content),
      }))
      .map(({ path, data, content }) => ({
        data,
        content: parseMarkdown(path, content),
      }))
  }

  getStaticPostsPaths() {
    const isProd = process.env.NODE_ENV === 'production'
    return globSync(path.join('content/**/*.md'))
      .map((p) => ({ path: p, content: fs.readFileSync(p, 'utf8') }))
      .map(({ path, content }) => ({ path, content: fromFrontMatter(content) }))
      .filter(({ content }) => !isProd || content.test !== true)
  }
}
