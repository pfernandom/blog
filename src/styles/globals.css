html,
body {
  padding: 0;
  margin: 0;
  font-family: -apple-system, BlinkMacSystemFont, Segoe UI, Roboto, Oxygen,
    Ubuntu, Cantarell, Fira Sans, Droid Sans, Helvetica Neue, sans-serif;
  /* font: 100%/1.75 "Avenir Next", "Helvetica Neue", "Segoe UI", "Helvetica",
    "Arial", sans-serif; */
}

footer {
  text-align: center;
}

code {
  background: #f5f2f0;
  padding: 0em 0.5em;
  display: inline-block;
  font-family: Roboto;
  font-size: 1.1em;
}

@media (prefers-color-scheme: dark) {
  html {
    color-scheme: dark;
  }
  body {
    color: white;
    background: black;
  }
  code {
    background-color: grey;
  }
  pre {
    filter: brightness(0.95);
  }
}

h1,
h2,
h3,
h4 {
  font-family: Montserrat, sans-serif;
  font-weight: 500;
  text-rendering: optimizeLegibility;
  line-height: 1.1;
}

h2,
h3,
h4 {
  margin-top: 2em;
}

h1 {
  font-size: 2em;
  color: var(---heading-color);
}

h2 {
  font-size: 1.5em;
}

h2,
h3 {
  color: var(---heading-color);
  filter: brightness(0.8);
}

a:hover {
  text-decoration: underline;
}

p {
  margin-bottom: 1.5em;
}

.layout-content {
  max-width: 900px;
  margin: 1em auto;
  padding: 0 2em;
}

a {
  color: var(---heading-color);
}

a:hover {
  transform: scale(1.05);
  display: inline-block;
}

@import url(./blog.css);
@import url(./instagram.css);
