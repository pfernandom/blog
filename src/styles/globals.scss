html,
body {
  padding: 0;
  margin: 0;
  font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen,
    Ubuntu, Cantarell, 'Fira Sans', 'Droid Sans', 'Helvetica Neue', sans-serif;

  /* font: 100%/1.75 "Avenir Next", "Helvetica Neue", "Segoe UI", "Helvetica",
    "Arial", sans-serif; */
}

footer {
  text-align: center;
}
/* 
code {
  background: #f5f2f0;
  padding: 0 0.5em;
  display: inline-block;
  font-family: Roboto, sans-serif;
  font-size: 1.1em;
} */

:not(pre) code {
  margin: 1px;
  padding-left: 1px;
  padding-right: 1px;
}

@media (prefers-color-scheme: dark) {
  html {
    color-scheme: dark;
  }

  body {
    color: white;
    background: black;
  }

  code {
    background-color: grey;
  }

  pre {
    filter: brightness(0.95);
  }
}

h1,
h2,
h3,
h4 {
  font-family: Montserrat, sans-serif;
  font-weight: 500;
  text-rendering: optimizelegibility;
  line-height: 1.1;
}

h2,
h3,
h4 {
  margin-top: 2em;
}

h1 {
  font-size: 2em;
  color: var(---heading-color);
}

h2 {
  font-size: 1.5em;
}

h2,
h3,
h4 {
  color: var(---heading-color);
  filter: brightness(0.8);
}

a:hover {
  text-decoration: underline;
}

p {
  margin-bottom: 1.5em;
}

.layout-content {
  max-width: 900px;
  margin: 1em auto;
  padding: 0 2em;
}

a {
  color: var(---heading-color);
  text-decoration: none;
}

a:hover {
  text-decoration: underline;
}

@import './blog.scss';
@import './instagram.scss';

blockquote {
  --background: #fff;
  --color: #00cc8f;
  --font-size: 1.2em;

  /* padding: 0.5em;
  background-color: gainsboro;
  font-family: Courier; */
  border-left: 3px solid var(--color);
  color: #1a1a1a;
  font-family: Georgia, Times, 'Times New Roman', serif;
  font-size: var(--font-size);
  font-style: italic;
  line-height: 1.8em;
  margin-block-start: 0;
  margin-block-end: 0;
  margin-left: 0;
  padding: 1em 2em;
  position: relative;
  transition: 0.2s border ease-in-out;
  z-index: 0;
  width: 80%;
}

blockquote p {
  margin: 0;
  margin-bottom: 0.5em;
}

blockquote::before {
  content: '';
  position: absolute;
  top: 50%;
  left: -4px;
  height: 2em;
  background-color: var(--background);
  width: 5px;
  margin-top: -1em;
}

blockquote::after {
  content: '\201C';
  position: absolute;
  top: 50%;
  left: -0.5em;
  color: var(--color);
  font-style: normal;
  font-size: 2em;
  font-weight: bold;
  line-height: 1.5em;
  text-align: center;
  text-indent: -2px;
  width: 1em;
  margin-top: -0.5em;
  transition: 0.2s all ease-in-out, 0.4s transform ease-in-out;
}

blockquote.active:after {
  transform: rotateY(360deg);
}

blockquote h3 {
  margin-top: 0;
  font-size: var(--font-size);
}
