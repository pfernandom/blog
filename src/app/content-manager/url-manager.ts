import { CONTENT_PATH } from 'blog_constants'
import { Post } from './manager'

const SOURCE_ROOT = 'src/app'
const CONTENT_ROOT = CONTENT_PATH
const MDX_EXTENSION = '/index.mdx'

export default class UrlManager {
  subBlogRegex: RegExp

  constructor() {
    this.subBlogRegex = new RegExp(`${CONTENT_ROOT}/(\\w+)/`)
  }

  getSubBlogName(postPath: string) {
    const res = this.subBlogRegex.exec(postPath)
    return res ? res[1] : 'root'
  }

  getImagesDir(postPath: string) {
    return postPath
      .replace(SOURCE_ROOT, 'opt_images')
      .replace(MDX_EXTENSION, '')
  }
  getSlugFromFilePath(path: string, post: Pick<Post, 'date' | 'legacy'>) {
    const date = new Date(post.date)
    const isLegacy = post.legacy
    const slug = path
      .replace(/(\d{1,2})_/gi, `${date.getMonth() + 1}/`)
      .replace(MDX_EXTENSION, '')
    if (isLegacy) {
      return slug.replace(CONTENT_ROOT, '').replace(/^[/]?\w+\//, 'blog/')
    }
    return slug.replace(CONTENT_ROOT, 'blog')
  }

  getSlugFromPostInfo(post: Post) {
    const date = new Date(post.date)

    return `blog/${date.getFullYear()}/${date.getMonth() + 1}/${post.slug}`
  }
}
