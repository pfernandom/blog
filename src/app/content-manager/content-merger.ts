import { getNextAndPrevSeries } from 'app/_components/blog-post/blog-post-series'
import urlGetterFactory from 'app/helpers/url-getter-factory'
import { ReactElement } from 'rehype-react/lib'
import { LocalContentManager, Post } from './manager'
import { StaticContentManager } from './static-content-manager'

interface PostRenderData {
  post: Post
  isPartOfSeries: boolean
  seriesPosts: Post[]
  prev: Post | null
  next: Post | null
  seoImages?: string[]
  pageLinks?: { next: string | null; prev: string | null }
}

interface MdxPostRenderData extends PostRenderData {
  postPath: string
}

interface MdPostRenderData extends PostRenderData {
  content: ReactElement
}

interface PostUnion {
  slug: string
  postInfo: Post
  isMDX: boolean
}

export default class ContentMerger {
  allPosts: PostUnion[]

  constructor() {
    const contentManager = new LocalContentManager()
    const staticContentManager = new StaticContentManager()
    this.allPosts = [
      ...contentManager.fetchAllSync().map((mdxContent) => ({
        slug: mdxContent.slug.replace('blog/', ''),
        postInfo: mdxContent.frontmatter,
        isMDX: true,
      })),
      ...staticContentManager
        .getStaticPostsPaths()
        .map((p) => p.content)
        .filter((postInfo) => postInfo.slug)
        .map((postInfo) => ({
          slug: (postInfo.slug as string).replace('blog/', '') as string,
          postInfo,
          isMDX: false,
        })),
    ].sort(
      (a, b) =>
        new Date(b.postInfo.date).getTime() -
        new Date(a.postInfo.date).getTime()
    )
  }

  isPost(slug: string) {
    return this.allPosts.find((post) => post.slug === slug) ? true : false
  }

  isMDXPost(slug: string): boolean {
    return this.allPosts.find((post) => post.slug === slug)?.isMDX ?? false
  }

  fetchMDXContent(slug: string): MdxPostRenderData | null {
    if (!this.isMDXPost(slug)) {
      return null
    }
    const host = process.env['SITE_URL'] ?? ''
    const contentManager = new LocalContentManager()
    const allPosts = contentManager.fetchAllSync()

    const postsAndPaths: Array<[Post, string]> = allPosts.map((post) => {
      if (!post.frontmatter.slug) {
        post.frontmatter.slug = post.slug
        // TODO: Clean things so this is not necessary
        post.frontmatter.slug = post.frontmatter.slug.replace('blog/', '')
      }
      return [post.frontmatter, post.postPath]
    })

    // TODO: Handle default case
    const postAndPath = postsAndPaths.find((post) => {
      return post[0].slug?.endsWith(slug)
    })

    if (!postAndPath) {
      return null
    }
    const [post, postPath] = postAndPath
    const posts = postsAndPaths.map(([post]) => post)

    const seriesPosts = posts.filter(
      (p) => p.series && p.series === post?.series
    )

    const getPageUrl = urlGetterFactory(host)
    const prevAndNext = getNextAndPrev(this.allPosts, post)
    const { prevInSeries, nextInSeries } = getNextAndPrevSeries(
      post,
      seriesPosts
    )
    const { prev, next } = prevAndNext

    const seoImages = [post.hero_image_original ?? null, post.hero_image]
      .filter((img) => img != null)
      .map((img) => getPageUrl(img as string, true))

    const isPartOfSeries = seriesPosts.length > 1
    const prevInSeriesSlug = prevInSeries[prevInSeries.length - 1]?.slug
    const nextInSeriesSlug = nextInSeries[0]?.slug
    const prevS =
      prevInSeries.length > 0 && prevInSeriesSlug
        ? getPageUrl(prevInSeriesSlug)
        : null
    const nextS =
      nextInSeries.length > 0 && nextInSeriesSlug
        ? getPageUrl(nextInSeriesSlug)
        : null

    return {
      post,
      isPartOfSeries,
      seriesPosts,
      prev,
      next,
      seoImages,
      postPath,
      pageLinks: { next: nextS, prev: prevS },
    }
  }

  fetchStaticContent(slug: string): MdPostRenderData | null {
    if (this.isMDXPost(slug)) {
      return null
    }
    const contentManager = new StaticContentManager()
    const components = contentManager.getStaticPosts()
    const pagePost = components.find((c) => {
      return c.data['slug']?.replace('blog/', '') === slug
    })

    const content = pagePost?.content
    const postData = pagePost

    if (content && postData && postData.data) {
      const prevAndNext = getNextAndPrev(this.allPosts, postData.data)
      const { prev, next } = prevAndNext
      return {
        content,
        post: postData.data,
        isPartOfSeries: false,
        seriesPosts: [],
        prev,
        next,
      }
    }
    return null
  }
}

function getNextAndPrev(posts: Array<PostUnion>, currentPost: Post) {
  const filtered = posts.filter((p) => p.postInfo.published)

  const postIndex = filtered.findIndex((p) => p.slug === currentPost?.slug)
  const prevIndex = postIndex - 1

  const nextIndex = postIndex + 1

  const prev = prevIndex >= 0 ? filtered[prevIndex].postInfo : null
  const next = nextIndex < filtered.length ? filtered[nextIndex].postInfo : null

  return { prev, next }
}
