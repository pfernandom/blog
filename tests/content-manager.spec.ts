import { LocalContentManager } from 'app/content-manager/manager'
import { describe, expect, test } from '@jest/globals'
import fs from 'fs'
import path from 'path'

describe('Content Manager', function () {
  test('Fetches all subblogs', async function () {
    const contentManager = new LocalContentManager()

    const subblogs = await contentManager.findSubBlogs()

    expect(subblogs).toMatchSnapshot()
  })

  test('Fetches all posts for a subblog', async function () {
    const contentManager = new LocalContentManager()

    const posts = await contentManager.fetchAll()

    expect(posts.map((post) => post.slug)).toMatchSnapshot()
    expect(posts.map((post) => post.postPath)).toMatchSnapshot()
    const heroImage = path.join('public', posts[0].frontmatter.hero_image)
    try {
      expect(fs.existsSync(heroImage)).toBe(true)
    } catch (err) {
      throw new Error(`${heroImage} does not exist:\n ${err}`)
    }
  })
})
