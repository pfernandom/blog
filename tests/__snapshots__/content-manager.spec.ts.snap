// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Content Manager Fetches all posts for a subblog 1`] = `
[
  "blog/2023/2/node_golang_wasm_p2",
  "blog/2023/2/node_golang_wasm",
  "blog/2022/11/qb-dropout-in-cnn",
  "blog/2022/11/builder-design-pattern",
  "blog/2022/11/building-a-heap",
  "blog/2022/10/pytorch-classification",
  "blog/2022/9/rust-problems",
  "blog/2022/9/flutter-code-gen-2",
  "blog/2022/9/markdown-to-html",
  "blog/2022/8/flutter-code-gen-1",
  "blog/2022/8/book",
  "blog/2022/8/java-futures",
  "blog/2022/7/dark-mode-css",
  "blog/2022/7/flutter-app-state",
  "blog/2019/7/array-fill",
  "blog/2019/6/html-chunking",
  "blog/1988/6/test_post",
]
`;

exports[`Content Manager Fetches all posts for a subblog 2`] = `
[
  "src/app/content/software/2023/2_node_golang_wasm_p2/index.mdx",
  "src/app/content/software/2023/1_node_golang_wasm/index.mdx",
  "src/app/content/software/2022/13_qb-dropout-in-cnn/index.mdx",
  "src/app/content/software/2022/12_builder-design-pattern/index.mdx",
  "src/app/content/software/2022/11_building-a-heap/index.mdx",
  "src/app/content/software/2022/10_pytorch-classification/index.mdx",
  "src/app/content/software/2022/9_rust-problems/index.mdx",
  "src/app/content/software/2022/8_flutter-code-gen-2/index.mdx",
  "src/app/content/software/2022/7_markdown-to-html/index.mdx",
  "src/app/content/software/2022/6_flutter-code-gen-1/index.mdx",
  "src/app/content/software/2022/5_book/index.mdx",
  "src/app/content/software/2022/4_java-futures/index.mdx",
  "src/app/content/software/2022/3_dark-mode-css/index.mdx",
  "src/app/content/software/2022/1_flutter-app-state/index.mdx",
  "src/app/content/software/2019/2_array-fill/index.mdx",
  "src/app/content/software/2019/1_html-chunking/index.mdx",
  "src/app/content/software/1988/0_test_post/index.mdx",
]
`;

exports[`Content Manager Fetches all subblogs 1`] = `
[
  {
    "basename": "src/app/content",
    "metadata": {
      "description": "A blog about coding and all things software-development",
      "title": "Software development blog",
    },
    "path": "src/app/content/software",
    "slug": "software",
  },
]
`;
